node/src/exchange/client.rs

```diff
@@ -232,25 +232,31 @@ where
         let mut headers = Vec::with_capacity(responses.len());
 
         for response in responses {
+            // Unmarshal and validate
             let header = response.to_extended_header()?;
             trace!("Header: {header:?}");
             headers.push(header);
         }
 
         headers.sort_unstable_by_key(|header| header.height());
 
-        // TODO: verify against Store
+        // NOTE: Verification against store is done in the syncer
         match (&request.data, headers.len()) {
             // Allow HEAD requests to have any height in their response
             (Some(Data::Origin(0)), 1) => {}
 
-            // Check if all requested heights exist
+            // Headers are already validated individualy in `to_extended_header`
+            // but we need to verify that they are indeed a valid chain.
             (Some(Data::Origin(start)), amount) if *start > 0 && amount > 0 => {
-                for (header, height) in headers.iter().zip(*start..*start + amount as u64) {
-                    if header.height().value() != height {
-                        return Err(ExchangeError::InvalidResponse);
-                    }
+                let first = headers.first().expect("empty headers vector");
+
+                if first.height().value() != *start {
+                    return Err(ExchangeError::InvalidResponse);
                 }
+
+                first
+                    .verify_adjacent_range(&headers[1..])
+                    .map_err(|_| ExchangeError::InvalidResponse)?;
```

Comment:
I'm still not sure about this

Reply:
After a discussion with Celestia team, the verification must only be done in `get_verified_headers_range` and Syncer will use that method